distributed/tests/test_worker_plugins.py

```diff
@@ -0,0 +1,68 @@
+from distributed.utils_test import gen_cluster
+from distributed import Worker
+
+
+class MyPlugin:
+    name = "MyPlugin"
+
+    def __init__(self, data):
+        self.data = data
+
+    def setup(self, worker):
+        assert isinstance(worker, Worker)
+        self.worker = worker
+        self.worker._my_plugin_status = "setup"
+        self.worker._my_plugin_data = self.data
+
+    def teardown(self, worker):
+        assert isinstance(worker, Worker)
+        self.worker._my_plugin_status = "teardown"
+
+
+@gen_cluster(client=True, ncores=[])
+def test_create_with_client(c, s):
+    yield c.register_worker_plugin(MyPlugin(123))
+
+    worker = Worker(s.address, loop=s.loop)
+    yield worker._start()
+    assert worker._my_plugin_status == "setup"
+    assert worker._my_plugin_data == 123
+
+    yield worker._close()
+    assert worker._my_plugin_status == "teardown"
+
+
+@gen_cluster(client=True, worker_kwargs={"plugins": [MyPlugin(5)]})
+def test_create_on_construction(c, s, a, b):
+    assert len(a.plugins) == len(b.plugins) == 1
+    assert a._my_plugin_status == "setup"
+    assert a._my_plugin_data == 5
+
+
+@gen_cluster(client=True, worker_kwargs={"plugins": [MyPlugin(5)]})
+def test_idempotence_with_name(c, s, a, b):
```

Comment:
Are the names of these two tests flipped (this one doesn't use plugin names, the other does?)

Reply:
I think that we're good here.  The `MyPlugin` class has a `name` attribute, which is what causes the idempotence here.

In the test below, we change the name intentionally.

Reply:
Ah, understood.